src/Microsoft.VisualStudio.Threading/JoinableTask.cs

```diff
@@ -297,20 +297,29 @@ internal SynchronizationContext ApplicableJobSyncContext
         {
             get
             {
-                using (this.Factory.Context.NoMessagePumpSynchronizationContext.Apply())
+                if (this.Factory.Context.IsOnMainThread)
                 {
-                    lock (this.owner.Context.SyncContextLock)
+                    if (this.mainThreadJobSyncContext == null)
                     {
-                        if (this.Factory.Context.IsOnMainThread)
+                        using (this.Factory.Context.NoMessagePumpSynchronizationContext.Apply())
                         {
-                            if (this.mainThreadJobSyncContext == null)
+                            lock (this.owner.Context.SyncContextLock)
                             {
-                                this.mainThreadJobSyncContext = new JoinableTaskSynchronizationContext(this, true);
+                                if (this.mainThreadJobSyncContext == null)
+                                {
+                                    this.mainThreadJobSyncContext = new JoinableTaskSynchronizationContext(this, true);
+                                }
                             }
-
-                            return this.mainThreadJobSyncContext;
                         }
-                        else
+                    }
+
+                    return this.mainThreadJobSyncContext;
+                }
+                else
+                {
+                    using (this.Factory.Context.NoMessagePumpSynchronizationContext.Apply())
```

Comment:
💭 It may be possible to fast-path a negative case here:

```
bool canSynchronouslyBlockThreadPool =
    (state & JoinableTaskFlags.StartedSynchronously) == JoinableTaskFlags.StartedSynchronously
    && (state & JoinableTaskFlags.StartedOnMainThread) != JoinableTaskFlags.StartedOnMainThread;
```

If the operation cannot synchronously block the thread pool per that condition, then you can simply return null.

Reply:
I've convinced myself that an unconditional fast path is safe here.